KomaMRIFiles/src/Sequence/Pulseq.jl

```diff
@@ -434,9 +434,42 @@ function read_seq(filename)
     for i = 1:length(blockEvents)
         seq += get_block(obj, i)
     end
-    # Final details
     # Remove dummy seq block at the start, Issue #203
     seq = seq[2:end]
+
+    # Add first and last Pulseq points
+    grad_prev_last = [0.0; 0.0; 0.0]
+    for bi in 1:length(seq)
+        for gi in 1:3
+            gr = seq.GR[gi, bi]
+            if seq.DUR[bi] > 0
+                if sum(abs.(gr.A)) == 0   # this is for no-gradient case
+                    grad_prev_last[gi] = 0.0
+                    continue
+                else
+                    if gr.A isa Vector # this is for the uniformly-shaped or time-shaped case
+                        if gr.delay > 0
+                            grad_prev_last[gi] = 0.0
+                        end
+                        seq.GR[gi, bi].first = grad_prev_last[gi]
+                        if gr.T isa Array # this is for time-shaped case (I assume it is the extended trapezoid case)
+                            seq.GR[gi, bi].last = gr.A[end]
+                        else
+                            odd_step1 = [seq.GR[gi, bi].first; 2 * gr.A]
+                            odd_step2 = odd_step1 .* (mod.(1:length(odd_step1), 2) * 2 .- 1)
+                            waveform_odd_rest = cumsum(odd_step2) .* (mod.(1:length(odd_step2), 2) * 2 .- 1)
+                            seq.GR[gi, bi].last = waveform_odd_rest[end]
+                        end
+                        grad_prev_last[gi] = seq.GR[gi, bi].last
+                    else    # this is for the trapedoid case
+                        grad_prev_last[gi] = 0.0
+                    end
+                end
+            end
+        end
+    end
```

Comment:
Put this inside a function `fix_first_last_grads!(seq)`. Ideally like this:

Put inside function

```julia
fix_grads_first_last!(seq::Sequence, fill_method::PulseqFirstLastFill<:FirstLastFillMethod)
```
Or something like that. Feel free to choose a name for PulseqDefault.

If we apply the type aliases here, we can broadcast the Vector{Grad} and have subfunctions for each alias.

```julia
function fix_grads_first_last!(seq::Sequence, fill_method::FirstLastFillMethod)
	grad_prev_last = zeros(3)
	for s in seq
	    grad_prev_last = fix_grad_first_last!.(s.GR, grad_prev_last, Ref(fill_method))
	end
end
...
fix_grad_first_last!(grad::NoGrad,               grad_prev_last, fill_method::PulseqFirstLastFill) = [0;0;0]
fix_grad_first_last!(grad::TrapezoidalGrad,      grad_prev_last, fill_method::PulseqFirstLastFill) = ...
fix_grad_first_last!(grad::UniformlySampledGrad, grad_prev_last, fill_method::PulseqFirstLastFill) = ...
fix_grad_first_last!(grad::TimeShapedGrad,       grad_prev_last, fill_method::PulseqFirstLastFill) = ...```

Reply:
I simply created the function without creating the filling method. Addressed in [1a989f5](https://github.com/JuliaHealth/KomaMRI.jl/pull/321/commits/1a989f5f236663ca3df7263532afceae6f3ee2ae). I also added a reference in #362

Reply:
This does not cover all the comments. Please dispatch on a filltype and put Pulseq's as the default.